ipaserver/plugins/cert.py

```diff
@@ -1046,6 +1054,84 @@ def _principal_name_matches_principal(name, principal_obj):
     return principal in principal_obj.get('krbprincipalname', [])
 
 
+def _validate_san_ips(generalnames):
+    """
+    Raise a ValidationError if the subjectAltName in a CSR includes
+    any IP address(es) that do not match a DNS name in the SAN.
+
+    """
+    san_dns_ips = set()
+    for name in generalnames:
+        if isinstance(name, cryptography.x509.general_name.DNSName):
+            san_dns_ips.update(_san_dnsname_ips(name.value))
+    for name in generalnames:
+        if isinstance(name, cryptography.x509.general_name.IPAddress):
+            if unicode(name.value) not in san_dns_ips:
+                raise errors.ValidationError(
+                    name='csr',
+                    error=_(
+                        "IP address in subjectAltName (%s) does not "
+                        "match any DNS name"
+                    ) % name.value
+                )
+
+
+def _san_dnsname_ips(dnsname, dnsname_is_cname=False):
+    """
+    Returns a set of IP addresses, managed by this IPa instance,
+    that correspond to the DNS name (from the subjectAltName).
+
+    """
+    fqdn = dnsutil.DNSName(dnsname).make_absolute()
+    if fqdn.__len__() < 4:
+        logger.debug("Skipping IPs for %s: hostname too short" % dnsname)
+        return ()
```

Comment:
Please return empty `list` or empty `set` (probably the latter because you return `set` elsewhere in this subroutine).
`tuple` is semantically invalid here.

Reply:
Will do.